yubikey-manager-qt: fix detecting inserted yubikey and mark as insecure

avdv opened the pull request:
Fixes #353582

The upload functionality was removed in an upstream dependency, so importing the types failed and caused an error.

See https://github.com/Yubico/yubikey-manager-qt/issues/364#issuecomment-2578318174

Also, let's mark it as insecure as suggested by @Pandapip1 since it is no longer actively developed / maintained.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

avdv committed:
yubikey-manager-qt: fix detecting inserted yubikey

Fixes #353582

The upload functionality was removed in an upstream dependency, so importing the
types failed and caused an error.

See https://github.com/Yubico/yubikey-manager-qt/issues/364#issuecomment-2578318174

Disasm commented on pkgs/tools/misc/yubikey-manager-qt/remove-upload-import.diff:
```diff
@@ -0,0 +1,15 @@
+     from ykman.device import list_all_devices, scan_devices
+-    from ykman.otp import (
+-    _PrepareUploadFailed as PrepareUploadFailed
+-    , _prepare_upload_key as prepare_upload_key, generate_static_pw)
```
`generate_static_pw` still exists it seems, and the import line below is valid since it's under a version check.

avdv replied:
Thank you, fixed.

Pandapip1 commented:
This appears to be unnecessary, as it seems to currently be working.

avdv commented:
> This appears to be unnecessary, as it seems to currently be working.

Which revision did you test? It definitely does not work for me without a patch. I tried 121aa118d13f01a7e5c5b5df93f85a28bac6a190.

avdv committed:
yubikey-manager-qt: mark as insecure

The project is not under development anymore, Yubico Authenticator is recommended instead.

Pandapip1 commented:
I'm not sure what I did back then. I've since installed the flutter version and removed the original yubikey manager. I assume you have the more up-to-date info

nbraud commented on pkgs/tools/misc/yubikey-manager-qt/remove-upload-import.diff:
```diff
@@ -0,0 +1,16 @@
+-    from ykman.otp import (
+-    _PrepareUploadFailed as PrepareUploadFailed
+-    , _prepare_upload_key as prepare_upload_key, generate_static_pw)
++    from ykman.otp import (generate_static_pw)
```
You can drop the unnecessary parenthesis
```suggestion
+    from ykman.otp import generate_static_pw
```

nbraud commented on pkgs/tools/misc/yubikey-manager-qt/default.nix:
```diff
@@ -85,6 +87,9 @@ mkDerivation rec {
     description = "Cross-platform application for configuring any YubiKey over all USB interfaces";
     homepage = "https://developers.yubico.com/yubikey-manager-qt/";
+    knownVulnerabilities = [
+      "The YubiKey Manager is no longer under active development, install `yubioath-flutter` instead."
```
Is the plan to remove `yubikey-manager-qt` from nixpkgs after 25.05?  If so, the message should be explicit about it.

SuperSandro2000 replied:
```suggestion
      "Yubico has announced the End of Life of YubiKey Manager QT. Upstream recommends yubioath-flutter to configure a YubiKey. yubikey-manager-qt will be dropped after NixOS 25.05."
```

SuperSandro2000 commented:
We might as well drop it at this point as it is now broken for a few months and unmaintained.

Pandapip1 commented:
Agreed, but we should at least merge this PR.

SuperSandro2000 reviewed (approved):
Testing this from nixpkgs is not enough, you need to rebuild your system with it andt hen it works
